File: 1339.maximum-product-of-splitted-binary-tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

const MOD = 1_000_000_007

func maxProduct(root *TreeNode) int {
    var allSums []int
    var calculateSums func(*TreeNode) int

    calculateSums = func(root *TreeNode) int {
        if root == nil {
            return 0
        }
        sum := root.Val + calculateSums(root.Left) + calculateSums(root.Right)
        allSums = append(allSums, sum)
        return sum
    }

    totalSum := calculateSums(root)
    maxProduct := 0
    
    for _, sum := range allSums {
        product := sum * (totalSum - sum)
        maxProduct = max(maxProduct, product)
    }

    return maxProduct % MOD
}

File: 1339.maximum-product-of-splitted-binary-tree/solution.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
const val MOD = 1_000_000_007

class Solution {
    fun maxProduct(root: TreeNode?): Int {
        val allSums = mutableListOf<Long>()

        fun calculateSum(node: TreeNode?): Long {
            if (node == null) return 0L
            val currentSum = node.`val` + calculateSum(node.left) + calculateSum(node.right)
            allSums.add(currentSum)
            return currentSum
        }

        val totalSum = calculateSum(root)
        var maxProduct = 0L

        for (sum in allSums) {
            val product = sum * (totalSum - sum)
            if (product > maxProduct) maxProduct = product
        }

        return (maxProduct % MOD).toInt()
    }
}